src/Endpoint/Partition.php

```diff
@@ -0,0 +1,163 @@
+    }
+
+    public function offsetExists($offset)
+    {
+        return isset($this->definition[$offset]);
+    }
+
+    public function offsetUnset($offset)
+    {
+        unset($this->definition[$offset]);
+    }
+
+    /**
+     * Determine if this partition contains the provided region.
+     *
+     * @param string $region
+     *
+     * @return bool
+     */
+    public function matchesRegion($region)
+    {
+        if (isset($this->definition['regionRegex'])) {
+            return (bool) preg_match(
+                "@{$this->definition['regionRegex']}@",
+                $region
+            );
+        }
+
+        return isset($this->definition['regions'][$region]);
+    }
+
+    public function getRegionsForService($service)
+    {
+        if (isset($this->definition['services'][$service]['isRegionalized'])
+            && false === $this->definition['services'][$service]['isRegionalized']
+        ) {
+            return [$this->definition['services'][$service]['partitionEndpoint']];
+        }
+
+        if (isset($this->definition['services'][$service]['endpoints'])) {
+            return array_keys(
+                $this->definition['services'][$service]['endpoints']
+            );
+        }
+
+        return [];
+    }
+
+    public function __invoke(array $args = [])
+    {
+        $service = isset($args['service']) ? $args['service'] : '';
+        $region = isset($args['region']) ? $args['region'] : '';
+        $scheme = isset($args['scheme']) ? $args['scheme'] : 'https';
+        $data = $this->getEndpointData($service, $region);
+
+        return [
+            'endpoint' => "{$scheme}://" . $this->formatEndpoint(
+                    isset($data['hostname']) ? $data['hostname'] : '',
+                    $service,
+                    $region
+                ),
+            'signatureVersion' => $this->getSignatureVersion($data),
+            'signingRegion' => isset($data['credentialScope']['region'])
+                ? $data['credentialScope']['region']
+                : $region,
+            'signingName' => isset($data['credentialScope']['service'])
+                ? $data['credentialScope']['service']
+                : $service,
+        ];
+    }
+
+    private function getEndpointData($service, $region)
+    {
+
+        $resolvedRegion = $this->resolveRegion($service, $region);
+        $data = isset($this->definition['services'][$service]['endpoints'][$resolvedRegion])
+            ? $this->definition['services'][$service]['endpoints'][$resolvedRegion]
+            : [];
+        $data += isset($this->definition['services'][$service]['defaults'])
+            ? $this->definition['services'][$service]['defaults']
+            : [];
+        $data += isset($this->definition['defaults'])
+            ? $this->definition['defaults']
+            : [];
+
+        return $data;
+    }
+
+    private function getSignatureVersion(array $data)
+    {
+        static $supportedBySdk = [
```

Comment:
This is good, but on a side note: I wonder if we should add "sdkPreferredSignatureVersion" or something similar to the config file to remove this from each implementing SDK.